### PR TITLE
Add hooks for taggedLogger in the runtime.

### LINKED ISSUE #5560: Move handling of tagged telemetry properties to host layer
## Work Item
**Describe the outcome you expect**
Follow-on to https://github.com/microsoft/FluidFramework/pull/5308.  Handling of tagged telemetry properties belongs in the host layer, where logs may be uploaded via a telemetry system.

**Approach**
Update `ITelemetryBaseLogger.send` to take `ITaggableTelemetryProperties` instead of `ITelemetryProperties`.
> DONE

See comment below for detailed plan:  https://github.com/microsoft/FluidFramework/issues/5560#issuecomment-867843015

**Open questions**
Consider whether all our logging APIs should speak in `ITaggableTelemetryProperties`, not just error logging - probably so.
 > YES.  Done.

**Additional context**
See https://github.com/microsoft/FluidFramework/pull/5308/ for introduction of the tagging

Pre-requisite is #5561 

**Acceptance criteria**
Hosts are able to handle tagged telemetry data.  Follow-up change will be to remove our handling of tags in telemetry-utils/src/logger.ts

### PR DESCRIPTION
**Pull request by ksbrar**
Last step of https://github.com/microsoft/FluidFramework/issues/5560. Creates a tagged-logger adapter class (that can wrap vanilla loggers that do not handle tags) and adds runtime handling for taggedLogger property of IContainerContext.

### CODE REVIEW

File: packages/loader/container-loader/src/containerContext.ts [hunk 1]
```diff
@@ -86,7 +86,9 @@ export class ContainerContext implements IContainerContext {
         return context;
     }
 
+    /** @deprecated See IContainerContext for more details. */
     public readonly logger: ITelemetryLogger;
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/containerContext.ts:90]: make this a getter that only return the taggedLogger

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 2]
```diff
@@ -533,7 +534,9 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         containerScope: IFluidObject = context.scope,
         existing?: boolean,
     ): Promise<ContainerRuntime> {
-        const logger = ChildLogger.create(context.logger, undefined, {
+        // If taggedLogger exists, use it. Otherwise, wrap the vanilla logger:
+        const passLogger = context.taggedLogger ? context.taggedLogger : new TaggedLoggerAdapter(context.logger);
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/containerRuntime.ts:538]:
```suggestion
        const passLogger = context.taggedLogger  ?? new TaggedLoggerAdapter(context.logger);
```

File: api-report/driver-utils.api.md [hunk 3]
```diff
@@ -113,7 +113,7 @@ export function combineAppAndProtocolSummary(appSummary: ISummaryTree, protocolS
 export function configurableUrlResolver(resolversList: IUrlResolver[], request: IRequest): Promise<IResolvedUrl | undefined>;
 
 // @public (undocumented)
-export function createGenericNetworkError(errorMessage: string, canRetry: boolean, retryAfterMs?: number, props?: ITelemetryProperties): GenericNetworkError | ThrottlingError;
+export function createGenericNetworkError(errorMessage: string, canRetry: boolean, retryAfterMs?: number, props?: ITelemetryProperties): ThrottlingError | GenericNetworkError;
```
Comment (markfields) [inline api-report/driver-utils.api.md:116]:
```suggestion
export function createGenericNetworkError(errorMessage: string, canRetry: boolean, retryAfterMs?: number, props?: ITelemetryProperties): GenericNetworkError | ThrottlingError;
```

File: BREAKING.md [hunk 4]
```diff
@@ -32,6 +32,7 @@ The `webpack-fluid-loader` utility was previously available from a package named
 - [Property removed from ContainerRuntime class](#Property-removed-from-the-ContainerRuntime-class)
 - [attach() should only be called once](#attach-should-only-be-called-once)
 - [Loader access in data stores is removed](#loader-access-in-data-stores-is-removed)
+- [Logger without tag support marked deprecated in ContainerContext](#logger-without-tag-support-marked-deprecated-in-containercontext)
```
Comment (markfields) [inline BREAKING.md:35]: This happened in 0.44, not 0.45?  At any rate, you need a breaking change comment for 0.45 saying that this is actually happening (what you say below: "loggers passed to ContainerContext will need to support tagged events.").

File: BREAKING.md [hunk 5]
```diff
@@ -42,6 +43,9 @@ The `webpack-fluid-loader` utility was previously available from a package named
 ### Loader access in data stores is removed
 Following the deprecation warning [Loader in data stores deprecated](#loader-in-data-stores-deprecated), the associated APIs have now been removed.  In addition to the original deprecation notes, users will automatically have an `ILoader` available on the container scope object as the `ILoader` property if the container was created through a `Loader`.
 
+### Logger without tag support marked deprecated in ContainerContext
+The `logger` property of `ContainerContext` has been marked deprecated, and future versions of the loader will enforce the use of `taggedLogger` - that is, loggers passed to ContainerContext will need to support tagged events.
```
Comment (markfields) [inline BREAKING.md:47]:
```suggestion
The `logger` property of `ContainerContext` has been marked deprecated, and future versions of the loader will use `taggedLogger` - that is, loggers passed to ContainerContext will need to support tagged events.
```

File: packages/loader/container-loader/src/containerContext.ts [hunk 6]
```diff
@@ -86,7 +86,18 @@ export class ContainerContext implements IContainerContext {
...
+    public readonly taggedLogger: ITelemetryLogger;
+
+    /**
+     * Subtlety: returns this.taggedLogger since vanilla this.logger is now deprecated. See IContainerContext for more
+     * details.
+    */
+    public get logger(): ITelemetryLogger {
+        return this.taggedLogger;
+    }
+    public set logger(telemetryLogger: ITelemetryLogger) {
```
Comment (markfields) [inline packages/loader/container-loader/src/containerContext.ts:98]: Not necessary?

File: packages/loader/container-loader/src/containerContext.ts [hunk 7]
```diff
@@ -86,7 +86,18 @@ export class ContainerContext implements IContainerContext {
...
+        this.logger = telemetryLogger;
+    }
```
Comment (markfields) [inline packages/loader/container-loader/src/containerContext.ts:100]: Mark these as deprecated?

File: packages/utils/telemetry-utils/src/logger.ts [hunk 8]
```diff
@@ -218,6 +197,55 @@ export abstract class TelemetryLogger implements ITelemetryLogger {
...
+    public send(eventWithTagsMaybe: ITelemetryBaseEvent) {
+        const newEvent: ITelemetryBaseEvent = {
+            category: eventWithTagsMaybe.category,
+            eventName: eventWithTagsMaybe.eventName,
+        };
+        for (const key of Object.keys(getValidTelemetryProps(eventWithTagsMaybe, new Set()))) {
+            if (newEvent[key] !== undefined) {
+                // Don't overwrite existing properties on the event
+                continue;
+            }
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:217]: nit: I don't think this is necessary - newEvent starts empty except for category and eventName, and it's fine for them to get overwritten with the same value IMO.

File: packages/utils/telemetry-utils/src/logger.ts [hunk 9]
```diff
@@ -218,6 +197,55 @@ export abstract class TelemetryLogger implements ITelemetryLogger {
...
+                : { value: taggableProp, tag: undefined };
+            switch (tag) {
+                case undefined:
+                    // No tag means we can log plainly
+                    newEvent[key] = value;
+                    break;
+                case TelemetryDataTag.PackageData:
+                    // For Microsoft applications, PackageData is safe for now
+                    // (we don't load 3P code in 1P apps)
+                    // But this determination really belongs in the host layer
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:230]:
```suggestion
```

File: packages/utils/telemetry-utils/src/logger.ts [hunk 10]
```diff
@@ -218,6 +197,55 @@ export abstract class TelemetryLogger implements ITelemetryLogger {
...
+    public constructor(
+        private readonly logger: ITelemetryBaseLogger) {
+    }
+
+    public send(eventWithTagsMaybe: ITelemetryBaseEvent) {
+        const newEvent: ITelemetryBaseEvent = {
+            category: eventWithTagsMaybe.category,
+            eventName: eventWithTagsMaybe.eventName,
+        };
+        for (const key of Object.keys(getValidTelemetryProps(eventWithTagsMaybe, new Set()))) {
```
Comment (markfields) [inline packages/utils/telemetry-utils/src/logger.ts:213]: After thinking it over, I don't think we should call `getValidTelemetryProps` here.  It's overly paranoid - it would only kick in if someone used bad casting to shove an object onto an `ITelemetryProperties`.  And if they did and we _don't_ include this call, it'll still work out - `value` and `tag` will probably be `undefined` which is harmless, or even if they're set to something else, it's a bug and we don't need to guard against it.

That function (getValidTelemetryProps) was written for `LoggingError` where it's easy to add something to an error object needed at runtime, not thinking about the fact that it will also try to be logged (e.g. this happened with ODSP error facet codes). But here we're just copying from event to event and the type system should be good enough.

Comment (markfields) [reply packages/utils/telemetry-utils/src/logger.ts:213]: The other thought experiment that convinces me is:  Should we be calling `getValidTelemetryProps` in `TelemetryLogger.prepareEvent` or some similar place?  No, we will trust the type system and people to not go out of their way to shoot themselves in the foot.

**Review by markfields — COMMENTED** [review]
Looking good so far!  Left a couple suggestions for small tweaks.

As for the test failures - You should be able to cut/paste many of `prepareErrorObject`'s unit tests over to some new unit tests for `TaggedLoggerAdapter`.

Let's also pair for a few minutes to see if we can't get some compat tests going here to test the loader/runtime boundary.

File: BREAKING.md (hunk 4 above)
Comment (ksbrar) [reply BREAKING.md:35]: woops - yes this is 0.45

File: packages/utils/telemetry-utils/src/logger.ts (hunk 10 above)
Comment (ksbrar) [reply packages/utils/telemetry-utils/src/logger.ts:213]: Makes sense, not an adversarial scenario

File: packages/loader/container-loader/src/containerContext.ts [hunk 11]
```diff
@@ -173,6 +186,7 @@ export class ContainerContext implements IContainerContext {
 
     ) {
         this.logger = container.subLogger;
```
Comment (markfields) [inline packages/loader/container-loader/src/containerContext.ts:188]: nit: You can remove this line

File: packages/loader/container-loader/src/containerContext.ts [hunk 12]
```diff
@@ -86,7 +86,20 @@ export class ContainerContext implements IContainerContext {
...
+     * Subtlety: returns this.taggedLogger since vanilla this.logger is now deprecated. See IContainerContext for more
+     * details.
+    */
+    /** @deprecated See IContainerContext for more details. */
+    public get logger(): ITelemetryLogger {
+        return this.taggedLogger;
+    }
+    /** @deprecated See IContainerContext for more details. */
+    public set logger(telemetryLogger: ITelemetryLogger) {
+        this.logger = telemetryLogger;
```
Comment (markfields) [inline packages/loader/container-loader/src/containerContext.ts:101]: I think this is an infinite loop 😎

Comment (markfields) [reply packages/loader/container-loader/src/containerContext.ts:101]: You can just no-op in here I think...?  Or can we not have a setter since it was readonly?

**Review by markfields — APPROVED** [review]
🎂 yaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaay
